src/shogun/mathematics/graph/Array.h

```diff
@@ -0,0 +1,237 @@
+		 */
+		class Array
+		{
+		protected:
+			struct Protected
+			{
+			};
+			Array(Protected)
+			{
+			}
+
+			std::unique_ptr<LazyExpr> m_lazy_expr;
+			std::shared_ptr<Tensor> m_output_tensor;
+
+		public:
+			template <typename T>
+			Array(const T& input) : m_output_tensor(input)
+			{
+				m_lazy_expr = std::make_unique<LazyExpr>(m_output_tensor);
+			}
+
+			template <typename T>
+			Array(T&& input)
+			    : m_output_tensor(std::make_shared<Tensor>(std::move(input)))
+			{
+				m_lazy_expr = std::make_unique<LazyExpr>(m_output_tensor);
+			}
+
+			template <typename T>
+			static std::shared_ptr<Array> create_view(const T& input)
+			{
+				auto result = std::make_shared<Array>(Protected{});
+				result->m_output_tensor = Tensor::create_view(input);
+				result->m_lazy_expr =
+				    std::make_unique<LazyExpr>(result->m_output_tensor);
+				return result;
+			}
+
+			Array(std::unique_ptr<LazyExpr>&& expr)
+			    : m_lazy_expr(std::move(expr))
+			{
+			}
+
+			const Shape& get_shape() const
+			{
+				return m_lazy_expr->get_output()->get_shapes()[0];
+			}
+
+			const std::shared_ptr<NumberType>& get_type() const
+			{
+				return m_lazy_expr->get_output()->get_types()[0];
+			}
+
+			const std::unique_ptr<LazyExpr>& get_lazy_expression() const
+			{
+				return m_lazy_expr;
+			}
+
+			std::shared_ptr<Tensor> evaluate()
+			{
+				if (!m_output_tensor)
+					force_evaluate();
+				return m_output_tensor;
+			}
+
+		private:
+			void force_evaluate()
+			{
+				const auto& inputs = m_lazy_expr->get_inputs();
+				const auto& output_node = m_lazy_expr->get_output();
+
+				std::vector<std::shared_ptr<Tensor>> input_tensors;
+				std::vector<std::shared_ptr<node::Input>> input_nodes;
+
+				for (const auto& [node, tensor] : inputs)
+				{
+					input_nodes.push_back(node);
+					input_tensors.push_back(tensor);
+				}
+
+				// evaluate the graph
+				auto graph = std::make_shared<Graph>(
+				    input_nodes, std::vector{output_node});
+				// build with default backend
+				graph->build();
+
+				auto result = graph->evaluate(input_tensors);
+				// there should be only one output here
+				require(
+				    result.size() == 1,
+				    "Expected only one output Tensor in Array evaluation.");
+
+				m_output_tensor = result[0];
+			}
+		};
+
+		std::shared_ptr<Array> operator+(
+		    const std::shared_ptr<Array>& lhs,
+		    const std::shared_ptr<Array>& rhs)
+		{
+			auto expr = lhs->get_lazy_expression()->copy();
+
+			expr->bind<node::Add>(rhs->get_lazy_expression());
+			return std::make_shared<Array>(std::move(expr));
+		}
+
+		std::shared_ptr<Array> matmul(
+		    const std::shared_ptr<Array>& lhs,
+		    const std::shared_ptr<Array>& rhs, const bool transpose_a,
+		    const bool transpose_b)
+		{
+			auto expr = lhs->get_lazy_expression()->copy();
+			expr->bind<node::MatMul>(
+			    rhs->get_lazy_expression(), transpose_a, transpose_b);
+			return std::make_shared<Array>(std::move(expr));
+		}
```

Comment:
@vigsterkr is it worth coming up with a way to automatically expose nodes via this interface? I am thinking we could use ctags? Or it can just be done manually. This will have to be defined in a .cpp too, otherwise won't work with interfaces